### PR TITLE
Allow for generic auth headers with pystac + validators

### PR DESCRIPTION
**Pull request by john-dupuy**
The aim of this PR is to build off of https://github.com/stac-utils/pystac/pull/889 so that generic auth headers can be added to `stac-api-validator` and dependent requests that are made by `pystac`. 

- Adds an `--auth-headers` click option that can take a generic auth header dict as a string and attach it to dependent requests
- Modifies the `stac_io` of the pystac `Catalog` so that the headers will be used as the catalog gets "walked"
- Bumps the `pystac` version so that https://github.com/stac-utils/pystac/pull/889 is included
- Bumps `typeguard` to `3.0.2`

TODO:
- [x] documentation
- [x] tests

### CODE REVIEW

**Comment by john-dupuy** [comment]
@gadomski or @philvarner - could one of you take a look? First contribution to this repo so please let me know if anything is missing.

**Comment by philvarner** [comment]
> @gadomski or @philvarner - could one of you take a look? First contribution to this repo so please let me know if anything is missing.

Thanks, I'll take a look this week.

File: tests/test_main.py
Comment (john-dupuy) [inline tests/test_main.py:25]: I couldn't figure out a better workaround for this. It seems that running this test with the `pytest-typeguard` plugin does some funny stuff to imports which breaks my mock below. 

I believe it's coming from the fact that it does [`install_import_hook`](https://github.com/agronholm/typeguard/blob/db7771024eb2f6eba03c3954f8abe0f03bded3c2/src/typeguard/_pytest_plugin.py#L71) which makes the mock patch path no longer valid, but I couldn't find a way to fix it. The test works when running pytest without the `--typeguard-packages` option.

File: tests/test_main.py
Comment (gadomski) [inline tests/test_main.py:48]: Since most of this PR's changes are to `validations.py`, I'd recommend adding tests that exercise the new behavior for the following functions:

- `get_catalog` (a new "public" function)
- `retrieve`
- `validate_api`
- `validate_browsable`

@philvarner may have a preferred test layout, but in liu of that, they could live in a `test_validations.py`.

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:372]: What's the reason for this change? It _looks_ like it's not changing behavior, but maybe I'm missing something.

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:530]: `auth_headers` is a misleading name for this argument, since it's updating the `headers`. I recommend changing the argument name to `headers`.

File: src/stac_api_validator/__main__.py
Comment (gadomski) [inline src/stac_api_validator/__main__.py:69]: Making this argument literally Python code is a little strange to me. I'd recommend using the same format as curl, e.g. (from their docs):

```
curl -H "X-First-Name: Joe" https://example.com
```

So, for your example, the expected string would be `"Authorization: api-key <api-key>"`.

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:282]: Why limit this copy to `Authorization`? Any reason not to copy all headers?

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:602]: Same question as in `get_catalog`, is there any reason to limit this copy to the `Authorization` header? Why not copy all headers?

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:749]: Where does `collection_id` come from? It's not part of the spec.

**Review by gadomski — CHANGES_REQUESTED** [review]

File: src/stac_api_validator/__main__.py
Comment (philvarner) [reply src/stac_api_validator/__main__.py:69]: I'd prefer for this to be `--auth-header` and a click "multiple" field, where each header specified is the HTTP format style that curl uses (as Pete said)

File: src/stac_api_validator/validations.py
Comment (john-dupuy) [reply src/stac_api_validator/validations.py:282]: The session has an `Accept-Encoding` header that pystac doesn't like. An example:
```
{'User-Agent': 'python-requests/2.28.2', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'api-key <api-key>'}
```
Results in an error:
```
- [Core] Error while traversing Catalog with pystac: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte 
```

If I change the `Accept-Encoding` header to `*` it works fine. I will update it so that this is the case.

File: src/stac_api_validator/validations.py
Comment (gadomski) [inline src/stac_api_validator/validations.py:305]: Is this line related to the authorization at all? Does it need to be set even if the authorization headers aren't set?

File: tests/test_main.py
Comment (gadomski) [reply tests/test_main.py:25]: Fine by me, I haven't used **typeguard** before so am unfamiliar with how it might interact w/ a running system, will defer to @philvarner on his preferred resolution.

File: src/stac_api_validator/validations.py
Comment (gadomski) [reply src/stac_api_validator/validations.py]: Oh I missed this response when I was dong my latest review. It still strikes me as strange that you need to set `Accept-Encoding` _only_ when setting authorization. It seems as though we should try to avoid mucking with `Accept-Encoding` if at all possible.

Comment (john-dupuy) [reply src/stac_api_validator/validations.py]: Oh it's not only when setting authorization, the same thing would be true even if we were copying over the headers from the session onto pystac's StacIO object, but not setting the `Authorization` header.

It could be a quirk of Planet's STAC API (which besides unit tests, I'm using to test this), let me try with a public API.

File: src/stac_api_validator/validations.py
Comment (john-dupuy) [reply src/stac_api_validator/validations.py:305]: I'm going to resolve this comment so we can continue to discuss in https://github.com/stac-utils/stac-api-validator/pull/316#discussion_r1179260335

File: src/stac_api_validator/validations.py
Comment (gadomski) [reply src/stac_api_validator/validations.py]: Could we just

```python
stac_io.headers.update({"Authorization": str(r_session.headers["Authorization"])})
```

so we don't touch the other stuff?

Comment (john-dupuy) [reply src/stac_api_validator/validations.py]: Since we made the argument generic (`-H/--headers`) and are now allowing for any custom header to be sent, perhaps we should just copy over the custom ones to pystac?

Comment (john-dupuy) [reply src/stac_api_validator/validations.py]: > It could be a quirk of Planet's STAC API (which besides unit tests, I'm using to test this), let me try with a public API

Seems to occur for microsoft planetary computer when that `Accept-Encoding: *` is commented out:
```console
 $ stac-api-validator --root-url https://planetarycomputer.microsoft.com/api/stac/v1 --conformance core --headers "Authorization: api-key fake-api-key" --log-level DEBUG                                                                 
...
Errors:
- [Core] Error while traversing Catalog with pystac: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte 
```

Comment (john-dupuy) [reply src/stac_api_validator/validations.py]: @gadomski following up on 
> Since we made the argument generic (-H/--headers) and are now allowing for any custom header to be sent, perhaps we should just copy over the custom ones to pystac?

Should I go ahead with this implementation?

Comment (gadomski) [reply src/stac_api_validator/validations.py]: Sure!

**Comment by jharrisonSV** [comment]
Any movement on this 👀

**Review by gadomski — APPROVED** [review]

**Comment by gadomski** [comment]
Note that this PR failed on **main** (probably not this PR's fault), so I reverted it in #391 and will re-open as a new PR so we can fix CI before merging.

**Pull request merged**
